zdm/survey.py

```diff
@@ -80,40 +82,242 @@ def __init__(self, state, survey_name:str,
+        #minval = np.min([self.wlogmean - self.maxsigma*self.wlogsigma,
+        #                self.slogmean - self.maxsigma*self.slogsigma,
+        #                np.log10(self.WMin)])
+        minval = np.log10(self.WMin)
+        maxval = np.log10(self.WMax)
+        # I haven't decided yet where to put the value of 1000 internal bins
+        # in terms of parameters.
+        self.internal_logwvals = np.linspace(minval,maxval,self.NInternalBins)
+        
+        # initialise probability bins
+        if self.meta['WMETHOD'] == 3:
             # evaluate efficiencies at each redshift
-            self.NWbins = state.width.WNbins
             self.efficiencies = np.zeros([self.NWbins,self.NZ,self.NDM])
             self.wplist = np.zeros([self.NWbins,self.NZ])
-            self.wlist = np.zeros([self.NWbins,self.NZ])
             self.DMlist = np.zeros([self.NZ,self.NDM])
             self.mean_efficiencies = np.zeros([self.NZ,self.NDM])
             
+            if self.backproject:
+                self.pws = np.zeros([self.NZ,self.internal_logwvals.size,self.NWbins]) #[iz,:,:] = pw
+                self.ptaus = np.zeros([self.NZ,self.internal_logwvals.size,self.NWbins]) #[iz,:,:] = ptau
+            
             # we have a z-dependent scattering and width model
             for iz,z in enumerate(self.zvals):
-                pwidths,pprobs=make_widths(self, state, z=z)
-                _ = self.get_efficiency_from_wlist(pwidths,pprobs,
-                                        model=width_bias, edir=edir,iz=iz)
-        self.calc_max_dm()
-
+                self.make_widths(iz)
+                #_ = self.get_efficiency_from_wlist(self.wlist,self.wplist[:,iz],
+                #                        model=self.meta['WBIAS'], edir=self.edir, iz=iz)
+        else:
+            self.wplist = np.zeros([self.NWbins])
+            self.make_widths()
+            if self.backproject:
+                self.pws = np.zeros([self.internal_logwvals.size,self.NWbins]) #[iz,:,:] = pw
+                self.ptaus = np.zeros([self.internal_logwvals.size,self.NWbins]) #[iz,:,:] = ptau
+            #_ = self.get_efficiency_from_wlist(self.wlist,self.wplist,
+            #                            model=self.meta['WBIAS'], edir=self.edir, iz=None) 
+        self.do_efficiencies()
+        
+    def do_efficiencies(self):
+        """
+        Function to handle calculation of efficiencies.
+        Allows this to be called externally, e.g. if DM halo model changes
+        """
+        if self.meta['WMETHOD'] == 3:
+            for iz,z in enumerate(self.zvals):
+                _ = self.get_efficiency_from_wlist(self.wlist,self.wplist[:,iz],
+                                        model=self.meta['WBIAS'], edir=self.edir, iz=iz)
+        else:
+            _ = self.get_efficiency_from_wlist(self.wlist,self.wplist,
+                                        model=self.meta['WBIAS'], edir=self.edir, iz=None) 
+    def make_widths(self,iz=None):
```

Comment:
pretty intense code;  I hope it is right ;)

Reply:
Tis part was! Though I found some kinda-bugs somewhere else (well, barely, but it counts to me)